### PR TITLE
Add ability to prefetch in `cudf.pandas` and change default to managed pool

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR adds ability to prefetch in `cudf.pandas` based off of: https://github.com/rapidsai/rmm/pull/1608/

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:53]: Why is only “managed” prefetching? I think we were testing a prefetching managed _pool_. Worth checking before we merge.

Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py:53]: @GregoryKimball @vyasr Can you recommend on which all rmm modes need to have prefetching enabled?

Comment (mhaseeb123) [reply python/cudf/cudf/pandas/__init__.py:53]: Same thoughts as Bradley about the `managed_pool` having prefetch on sub-allocate.

Comment (bdice) [reply python/cudf/cudf/pandas/__init__.py:53]: I spoke with @galipremsagar and will pursue this fix.

Comment (bdice) [reply python/cudf/cudf/pandas/__init__.py:53]: Done in 7047b77366c93723de8a3c40fd3a2fbb16c05c19.

**Comment by GregoryKimball** [comment]
Hello @galipremsagar would you also please enable pylibcudf experimental prefetching? Ideally we can get these changes to the main branch by end-of-day.

**Comment by galipremsagar** [comment]
> Hello @galipremsagar would you also please enable pylibcudf experimental prefetching? Ideally we can get these changes to the main branch by end-of-day.

Done! This should be ready for review.

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:29]: os.getenv can’t return None, now that we provide a default value that is not None. Let’s refactor this now that the conditional and walrus operator are not needed.

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:77]: We only want to enable prefetching for managed and managed pool options. Otherwise it adds overhead for prefetching that isn’t usable without managed memory. I would initialize a variable `enable_prefetching = False`, set it to True for managed options, and conditionally enable prefetching based on its value.

Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Done 👍

File: python/cudf/cudf/pandas/__init__.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Done

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:37]: We could do this and remove the parts where we set it to True.
```suggestion
        enable_prefetching = "managed" in rmm_mode
```

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:69]: Let’s defer the import and put it one line lower, inside the ”if prefetching is enabled” block.

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:72]: I think the term for these is a “key.”
```suggestion
            for key in {
```

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:78]:
```suggestion
                pylibcudf.experimental.enable_prefetching(key)
```

**Review by bdice — APPROVED** [review]
A few small simplifications, then LGTM. I do not know if there are any other changes needed though.

**Review by mhaseeb123 — APPROVED** [review]

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:29]: What does "disable" do? Does it just not change the default mr at all? Does that have any meaningful difference from "cuda"? If not, can we remove "disable" as an option and just have "cuda" be the normal default behavior of using cudaMalloc (the default rmm mr)?

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:34]: this warnings doesn't seem accurate. We need some sort of early return in place in order to ignore the env var, otherwise we will process it regardless of what the current mr is.

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:64]: Can we move the `set_current_device_resource` calls outside the conditional? The only conditional part is the actual construction of `mr`, so if we default `mr` to `current_mr` we can unconditionally call `set_current_device_resource` after the conditional.

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:69]: No real need to make this a function-local import, is there? Unless it causes some (very surprising) circular import issues I would just import it at the top level of the file unconditionally. The module will be loaded by importing cudf anyway so it doesn't save us any import time.

**Review by vyasr — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/pandas/__init__.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Dropped `disable` and `cuda` will now act as disabling option.

File: python/cudf/cudf/pandas/__init__.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Done

File: python/cudf/cudf/pandas/__init__.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Done

File: python/cudf/cudf/pandas/__init__.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Introduced an early return.

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:48]: We already got the current MR earlier on line 32.
```suggestion
                current_mr,
```

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:69]: Might be a bit more explicit to put `new_mr = current_mr` before the conditional, then reassign `new_mr`. Then at the end you're not setting `current_mr` again. Just makes it easier and more obvious to the reader

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:45]:
```suggestion
    if rmm_mode == "pool":
```

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:67]:
```suggestion
    elif rmm_mode != "cuda":
```

File: python/cudf/cudf/pandas/__init__.py
Comment (vyasr) [inline python/cudf/cudf/pandas/__init__.py:68]: Cool things you can do with f-strings!
```suggestion
        raise ValueError(f"Unsupported {rmm mode=}")
```

**Review by vyasr — APPROVED** [review]
Some suggestions, but I don't need to review again.

File: python/cudf/cudf/pandas/__init__.py
Comment (bdice) [reply python/cudf/cudf/pandas/__init__.py:68]: This needs an underscore and not a space, I think.
```suggestion
        raise ValueError(f"Unsupported {rmm_mode=}")
```

Comment (vyasr) [reply python/cudf/cudf/pandas/__init__.py]: Oh good catch thanks.

File: python/cudf/cudf/pandas/__init__.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: Done 👍

**Review by harrism — APPROVED** [review]

**Review by vyasr — APPROVED** [review]
Approving again, this time on behalf of ci-codeowners due to the parallelism reduction in CI testing required to avoid issues with the default managed pool mr used in cudf.pandas when testing in parallel with pytest-xdist.

File: python/cudf/cudf/pandas/scripts/run-pandas-tests.sh
Comment (galipremsagar) [inline python/cudf/cudf/pandas/scripts/run-pandas-tests.sh:140]:
```suggestion
PANDAS_CI="1" timeout 60m python -m pytest -p cudf.pandas \
```

**Pull request merged**
